compiler/passes/convert-uast.cpp

```diff
@@ -387,88 +390,97 @@ struct Converter {
+                return new SymExpr(sym);
               }
             }
+          }
 
-            // figure out if it is field access
-            bool isFieldAccess = false;
-            const uast::Formal* parentMethodThis = nullptr;
-            if (parsing::idIsField(context, id)) {
-              // are we in a method? if so, what is the 'this' formal?
-              for (auto it = symStack.rbegin(); it != symStack.rend(); ++it) {
-                if (it->ast != nullptr) {
-                  if (auto inFn = it->ast->toFunction()) {
-                    if (inFn->isMethod()) {
-                      parentMethodThis = inFn->thisFormal();
-                      isFieldAccess = true;
-                      break;
-                    }
+          // figure out if it is field access
+          bool isFieldAccess = false;
+          const uast::Formal* parentMethodThis = nullptr;
+          Symbol* parentMethodConvertedThis = nullptr;
+          if (parsing::idIsField(context, id)) {
+            if (inForwardingMethod != nullptr) {
+              parentMethodConvertedThis = inForwardingMethod->_this;
+              isFieldAccess = true;
+            }
+
+            // are we in a method? if so, what is the 'this' formal?
+            for (auto it = symStack.rbegin(); it != symStack.rend(); ++it) {
+              if (it->ast != nullptr) {
+                if (auto inFn = it->ast->toFunction()) {
+                  if (inFn->isMethod()) {
+                    parentMethodThis = inFn->thisFormal();
+                    isFieldAccess = true;
+                    break;
                   }
                 }
               }
             }
+          }
 
-            // handle field access when only scope resolving
-            if (!shouldResolve(node) && isFieldAccess) {
-              // if we are just scope resolving, convert field
-              // access to this.field using a string literal to
-              // match production scope resolve
-              Symbol* thisSym = findConvertedSym(parentMethodThis->id());
-              INT_ASSERT(thisSym != nullptr);
-              auto ast = parsing::idToAst(context, id);
-              INT_ASSERT(ast && ast->isVariable());
-              auto var = ast->toVariable();
-              auto str = new_CStringSymbol(var->name().c_str());
-              CallExpr* ret = new CallExpr(".", thisSym, str);
-              return ret;
+          // handle field access when only scope resolving
+          if (!shouldResolve(node) && isFieldAccess) {
+            // if we are just scope resolving, convert field
+            // access to this.field using a string literal to
+            // match production scope resolve
+            Symbol* thisSym = nullptr;
+            if (parentMethodConvertedThis != nullptr) {
+              thisSym = parentMethodConvertedThis;
+            } else {
+              INT_ASSERT(parentMethodThis);
+              thisSym = findConvertedSym(parentMethodThis->id());
```

Comment:
I believe this is incorrect / insufficient. It is indeed possible to have a nested record / method _inside_ of a forwarding declaration. You just need a lambda (or an FCF, probably)!
```Chapel
proc string.method() {
    writeln("Called on a string: ", this);
}

proc int.method() {
    writeln("Called on an int: ", this);
}

record GlobalRecord {
    var x = "hello";
    forwarding ((lambda (){
        record LocalRecord {
            var x = 42;

            proc methodThatUsesThis() {
                return x;
            }
        }
        var localRecord: LocalRecord;
        return localRecord.methodThatUsesThis();
    })());
}
var r: GlobalRecord;
r.method();
```

How the above code should work: `methodThatUseThis` should return 42 from the lambda, and thus from the lambda call; thus, the program should print 42. This is indeed what the production compiler does:
```
➜  chapel git:(30b625e4bb) chpl localrecord.chpl
➜  chapel git:(30b625e4bb) ./localrecord
Called on an int: 42
```
On the other hand, because this code prefers the `parentMethodConvertedThis` to a regular method `this`, I would expect it to use the outer `x`, which is a `string`; it would be returned from `methodThatUsesThis`, and from the lambda, and to the forwarding expression. Thus, the string method would be called. And indeed:
```
➜  chapel git:(30b625e4bb) chpl localrecord.chpl --dyno
➜  chapel git:(30b625e4bb) ./localrecord
Called on a string: hello
```

Might I suggest using a stack of 'this anchors', which you modify whenever you enter / exit either a method proc or a forwarding decl? It would also help avoid searching upwards in a symStack for a method.

Presumably, this current limitation can also be problematic if you use a lambda, with a record, with a forwarding statement (which will overwrite the value of `inForwardingMethod` to something nonsensical!

Reply:
Wow, I'm surprised that works in production. I think it would be OK to make it an error. But, the proposal of having a stack of 'this anchors' also makes sense to me, and might simplify some of the related code, so I will try that.